### PR TITLE
provider/azurerm: add traffic manager resources

### PR DESCRIPTION
**Pull request by pmcatominey**
```
TF_ACC=1 go test ./builtin/providers/azurerm -v -run TestAccAzureRMTrafficManager -timeout 120m
=== RUN   TestAccAzureRMTrafficManagerEndpoint_basic
--- PASS: TestAccAzureRMTrafficManagerEndpoint_basic (163.31s)
=== RUN   TestAccAzureRMTrafficManagerEndpoint_basicDisableExternal
--- PASS: TestAccAzureRMTrafficManagerEndpoint_basicDisableExternal (179.61s)
=== RUN   TestAccAzureRMTrafficManagerEndpoint_updateWeight
--- PASS: TestAccAzureRMTrafficManagerEndpoint_updateWeight (245.83s)
=== RUN   TestAccAzureRMTrafficManagerEndpoint_updatePriority
--- PASS: TestAccAzureRMTrafficManagerEndpoint_updatePriority (101.39s)
=== RUN   TestAccAzureRMTrafficManagerEndpoint_nestedEndpoints
--- PASS: TestAccAzureRMTrafficManagerEndpoint_nestedEndpoints (71.72s)
=== RUN   TestAccAzureRMTrafficManagerProfile_weighted
--- PASS: TestAccAzureRMTrafficManagerProfile_weighted (78.26s)
=== RUN   TestAccAzureRMTrafficManagerProfile_performance
--- PASS: TestAccAzureRMTrafficManagerProfile_performance (85.58s)
=== RUN   TestAccAzureRMTrafficManagerProfile_priority
--- PASS: TestAccAzureRMTrafficManagerProfile_priority (86.91s)
=== RUN   TestAccAzureRMTrafficManagerProfile_withTags
--- PASS: TestAccAzureRMTrafficManagerProfile_withTags (103.82s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/azurerm	1116.543s
```

### CODE REVIEW

**Comment by stack72** [comment]
Hi @pmcatominey 

Thanks so much for the PR here - will run the tests and have a look through :)

P.

File: builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+		d.SetId("")
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("Error making Read request on TrafficManager Endpoint %s: %s", name, err)
+	}
+	endpoint := *resp.Properties
+
+	d.Set("name", resp.Name)
+	d.Set("endpoint_status", *endpoint.EndpointStatus)
```
Comment (stack72) [inline builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go]: We don't need to dereference before passing into d.Set - infact, it's safer if we don't, it takes care of the check

File: builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+
+	if endpoint.TargetResourceID != nil {
+		d.Set("target_resource_id", *endpoint.TargetResourceID)
+	}
+
+	if endpoint.Target != nil {
+		d.Set("target", *endpoint.Target)
+	}
+
+	if endpoint.Weight != nil {
```
Comment (stack72) [inline builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go]: As these are all strings and ints, we don't need to check if they are nil - d.Set will take care of this for us and dereference accordingly

File: builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+		mci64 := int64(minChildEndpoints)
+		endpointProps.MinChildEndpoints = &mci64
+	}
+
+	return &endpointProps
+}
+
+func validateAzureRMTrafficManagerEndpointType(i interface{}, k string) (s []string, errors []error) {
+	valid := map[string]struct{}{
+		"Microsoft.Network/TrafficManagerProfiles/azureEndpoints":    struct{}{},
```
Comment (stack72) [inline builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go]: Do we really need these entire strings? Doesn't the API add these for us if we pass in `azureEndpoints`, `externalEndpoints` or `nestedEndpoints`?

We haven't had to pass type before so wondering what the path of least resistance is here

Comment (pmcatominey) [inline builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go]: I was concerned that Microsoft was planning to allow other Endpoint types from different providers but discovered the short name is used in Resource ID, I'll shorten them for the user.

File: builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go (hunk 2 above)
Comment (pmcatominey) [inline builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go]: Good to know, should the type conversions be removed too? e.g. `d.Set("weight", int(*endpoint.Weight))`

Comment (stack72) [inline builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go]: Are you converting the types or is it an int from the api being set to an int in the schema?

Comment (pmcatominey) [inline builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go]: The API returns them as `int64` but the type assertions in the create function expect them to be `int` from the schema.

Comment (stack72) [inline builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go]: that's fine :) You shouldn't have to do the type assertions either

Comment (pmcatominey) [inline builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go]: Really? I've got code like this when preparing to send the request:
```
 +	if weight := d.Get("weight").(int); weight != 0 {
 +		w64 := int64(weight)
 +		endpointProps.Weight = &w64
 +	}
```

Comment (stack72) [inline builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go]: Hey

So in AWS, I do this:

```
Update Func
if d.HasChange("port") {
		updateHealthCheck.Port = aws.Int64(int64(d.Get("port").(int)))
	}

Read func
d.Set("port", updated.Port)
```

Comment (pmcatominey) [inline builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go]: I'll test that out, cheers.

Comment (pmcatominey) [inline builtin/providers/azurerm/resource_arm_traffic_manager_endpoint.go]: Bit of a snag, for some of these values Get and Set are not being called as they are lists:

```

**Comment by pmcatominey** [comment]
Removed the dereferences for top level attributes, they remain in place for properties of sets as it was cleaner than checking types.

**Comment by stack72** [comment]
Hi @pmcatominey 

This LGTM now :) I am going to hold off merging this until we finish cutting RC4 though - so bare with me for a day or 2

```
% make testacc TEST=./builtin/providers/azurerm TESTARGS='-run=TestAccAzureRMTrafficManager'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
TF_ACC=1 go test ./builtin/providers/azurerm -v -run=TestAccAzureRMTrafficManager -timeout 120m
=== RUN   TestAccAzureRMTrafficManagerEndpoint_basic
--- PASS: TestAccAzureRMTrafficManagerEndpoint_basic (137.88s)
=== RUN   TestAccAzureRMTrafficManagerEndpoint_basicDisableExternal
--- PASS: TestAccAzureRMTrafficManagerEndpoint_basicDisableExternal (182.37s)
=== RUN   TestAccAzureRMTrafficManagerEndpoint_updateWeight
--- PASS: TestAccAzureRMTrafficManagerEndpoint_updateWeight (167.46s)
=== RUN   TestAccAzureRMTrafficManagerEndpoint_updatePriority
--- PASS: TestAccAzureRMTrafficManagerEndpoint_updatePriority (122.34s)
=== RUN   TestAccAzureRMTrafficManagerEndpoint_nestedEndpoints
--- PASS: TestAccAzureRMTrafficManagerEndpoint_nestedEndpoints (98.02s)
=== RUN   TestAccAzureRMTrafficManagerProfile_weighted
--- PASS: TestAccAzureRMTrafficManagerProfile_weighted (75.61s)
=== RUN   TestAccAzureRMTrafficManagerProfile_performance
--- PASS: TestAccAzureRMTrafficManagerProfile_performance (88.42s)
=== RUN   TestAccAzureRMTrafficManagerProfile_priority
--- PASS: TestAccAzureRMTrafficManagerProfile_priority (93.49s)
=== RUN   TestAccAzureRMTrafficManagerProfile_withTags
--- PASS: TestAccAzureRMTrafficManagerProfile_withTags (114.20s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/azurerm	1079.803s
```

**Comment by pmcatominey** [comment]
Cheers

**Pull request merged**